Realtek 8723BS WiFi + BT module

This patch is a combination of the following patches:
https://github.com/armbian/build/blob/master/patch/kernel/sunxi-current/board-pine-h6-add-RTL8723.patch
https://github.com/torvalds/linux/commit/3d478d211b8e40deb609594f0676be205e8d89c1
https://forum.armbian.com/topic/13636-finally-got-bluetooth-working-on-pine64-h64-model-b/

Signed-off-by: AlexTMjugador <AlexTMjugador@users.noreply.github.com>
---
diff --git a/arch/arm64/boot/dts/allwinner/sun50i-h6-pine-h64-model-b.dts b/arch/arm64/boot/dts/allwinner/sun50i-h6-pine-h64-model-b.dts
index f4c8966a6..f9c892a10 100644
--- a/arch/arm64/boot/dts/allwinner/sun50i-h6-pine-h64-model-b.dts
+++ b/arch/arm64/boot/dts/allwinner/sun50i-h6-pine-h64-model-b.dts
@@ -9,9 +9,55 @@ / {
 	model = "Pine H64 model B";
 	compatible = "pine64,pine-h64-model-b", "allwinner,sun50i-h6";
 
+	wifi_pwrseq: wifi-pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		clocks = <&rtc 1>;
+		clock-names = "ext_clock";
+		reset-gpios = <&r_pio 1 3 GPIO_ACTIVE_LOW>; /* PM3 */
+		post-power-on-delay-ms = <200>;
+	};
+
 	/delete-node/ reg_gmac_3v3;
 };
 
+/*
+ * The Pine H64 model B has a Realtek 8723BS Wi-Fi + BT combo module.
+ * Wi-Fi is connected via SDIO on MMC1
+ */
+&mmc1 {
+	vmmc-supply = <&reg_cldo2>;
+	vqmmc-supply = <&reg_bldo3>;
+	mmc-pwrseq = <&wifi_pwrseq>;
+	bus-width = <4>;
+	non-removable;
+	status = "okay";
+
+	rtl8723bs: sdio-wifi@1 {
+		reg = <1>;
+		interrupt-parent = <&r_pio>;
+		interrupts = <1 0 IRQ_TYPE_LEVEL_LOW>; /* PM0 */
+		interrupt-names = "host-wake";
+	};
+};
+
+&reg_cldo2 {
+	/*
+	 * This regulator is connected with CLDO3.
+	 * Before the kernel can support synchronized
+	 * enable of coupled regulators, keep them
+	 * both always on as a ugly hack.
+	 */
+	regulator-always-on;
+};
+
+&reg_cldo3 {
+	/*
+	 * This regulator is connected with CLDO2.
+	 * See the comments for CLDO2.
+	 */
+	regulator-always-on;
+};
+
 &hdmi_connector {
 	/delete-property/ ddc-en-gpios;
 };
@@ -19,3 +65,19 @@ &hdmi_connector {
 &emac {
 	phy-supply = <&reg_aldo2>;
 };
+
+/* The BT submodule is connected to UART1 */
+&uart1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart1_pins>, <&uart1_rts_cts_pins>;
+	uart-has-rtscts;
+	status = "okay";
+
+	bluetooth {
+		compatible = "realtek,rtl8723bs-bt";
+		device-wake-gpios = <&r_pio 1 1 GPIO_ACTIVE_HIGH>; /* PM1 */
+		host-wake-gpios = <&r_pio 1 2 GPIO_ACTIVE_HIGH>; /* PM2 */
+		reset-gpios = <&r_pio 1 4 GPIO_ACTIVE_LOW>; /* PM4 */
+		post-power-on-delay-ms = <200>;
+	};
+};
